Detect blocking events

ghweber opened the pull request:
Add a detector for blocking events based on Tempest Extremes work.

kwrobot committed:
Cherry-pick teca_mesh_join from Amanda's branch

kwrobot committed:
Make teca_mesh_join available to Python

kwrobot committed:
Initial version of blocking event detection from TempestExtremes

Initial version of blocking event detection from TempestExtremes
consisting mainly of DetectBlobs (though limited to functions
needed for blocking event detection) and an application that
detects overlaps between time steps (first stage of StitchBlobs).

kwrobot committed:
Fix memory leak in teca_latitude_damper

kwrobot committed:
Add teca_latitude_range_filter

kwrobot committed:
Added methods to set periodicity to teca_cartesian_mesh_source

taobrienlbl commented on apps/teca_blocking_event_detect_blobs.in:
```diff
@@ -0,0 +1,378 @@
+t = input_coords['t']
+bounds = input_md['bounds']
+reader.set_periodic_in_x(True)
+reader.set_periodic_in_y(True) # TODO/FIXME: Is this really true for longitude?
```
periodic_in_y should be default False here and in the other places below

kwrobot committed:
Support for multiple teca_2d_component_area in one pipeline

Add ability to specify a prefix for generated metadata with component
2D area to support multiple filters in one pipeline without to compute
component area for multiple variables.

kwrobot committed:
Fix for .mcf input files. Further clean-up and refactoring.

kwrobot committed:
Add documentation

kwrobot committed:
Added tests for changes to 2d_component_area and new latitude_range_filter

kwrobot committed:
Fix errors found during writing tests.

kwrobot committed:
Added tests for blocking event detection.

Add tests for teca_blocking_event_detect_blobs and
teca_blocking_event_create_overlap_table. Do not verify blocking event
candidate detection using unique component ID but instead using a binary
mask (closer to how TempestExtremes saves data), as connected labeling in
TECA produces different results between CPU and GPU implementations.

kwrobot committed:
Add helper function for debugging

kwrobot committed:
Minor bug fix and improvements based on testing.

taobrienlbl commented on doc/rtd/applications.rst:
```diff
@@ -14,49 +14,56 @@ modules loaded) and to load the teca modulefiles as late as possible.  For
+| :ref:`teca_blocking_event_detect_blobs`    | Detect blocking event candidates as geopotential |
+|                                            | height anomalies above a size threshold.         |
++--------------------------------------------+--------------------------------------------------+
+| :ref:`teca_blocking_event_create_overlap_  | Create tracking table for blocking event         |
```
I think there's a missing backtick here - it renders with :ref: showing.  See https://teca.readthedocs.io/en/detect-blocking-events/applications.html

taobrienlbl reviewed (changes requested):
overall this looks awesome - great job @ghweber !

I'm going to request one additional, relatively significant change: a refactoring of the detect_blocking_event_detect_blobs app. The app itself is great, but the functionality for detecting blobs should be implemented as a an algorithm (e.g., as alg/teca_blocking_event_detect_blobs.py, with the app using the algorithm for detection. The reason for this is that we want to be able to use the blob detection algorithm within pipelines (e.g., for inline detection of blocking events in a ML weather simulation) and not just for reading from netCDF files.

taobrienlbl commented:
@ghweber - adding here an example of how we might use an algorithm-based version of the blob detection step: https://github.com/amandasd/teca_earth2studio/blob/1fe50ed82e4886780829c80b8b02b159dfa1e540/teca_tempest_tc_detect.py#L315 - we would use data passed from an on-GPU array as a TECA data source instead of the CF reader.  @amandasd can give insight here if needed

kwrobot committed:
Moved detect blobs into separate class.

kwrobot committed:
Reformat table in applications.rst to fix broken reference

kwrobot committed:
Move overlap table creation into separate class.

taobrienlbl commented:
hey @ghweber - the `test_blocking_event_create_overlap_table_app` test is failing: https://my.cdash.org/tests/319620247  It looks like there is a npy file that isn't found...does the SVN reference need updating?

taobrienlbl reviewed (approved):
Looks good - thank you @ghweber!  Once the issue with the test is fixed, this looks good to merge.

kwrobot committed:
Fix input argument validation

ghweber commented:
The failure was due to an incorrect consistency check added by AI (number of threads can be set to -1 meaning number of physical cores while the AI generated consistency check assumed >1) I corrected this and the tests pass now. I'll have another look at this tomorrow to see if I missed any other issues and then merge.

kwrobot committed:
Minor tweaks
